# Database : No.2 : Qdrant DataBase
-  URL : https://qdrant.tech/
- Qdrant Documentation = URL : https://qdrant.tech/documentation/advanced-tutorials/using-multivector-representations/
- Go to the "Tutorail" & click on -> "Advanced Retrieval"

- Let's go with the "Qdrant Database" :

- Simple Login with Google account. It is Basically a Shared Machine, if we want then we can goahead with dedecated machine Click on 
- Biling section and pay for it.
- Otherwise, if we just came for experimentation & learning purpose they are providing enough.
- Just click - "Create a Free Cluster" I will Name it -> "EuronVectorDB" & it is giving me -> Google cloude or AWS integration and 
- Free RAM and Disk & vCPUs etc.....
- SO, just provide the name and just click Create Cluster.
- It will provid me a "ApI Key" just save this API Key somewhere else, & if we see same place there is Python -> 
- Click on python inside there they have given , API key and Endpoint , they have attached my API, this details they have provided.
- So that I can hit there instance, we are able to hit there services. 

- Just copy this code from "Python section" & Paste below.
- Now we can see my instance is ready , my cluster is ready.

- Now we can see this all are the cluster resources they have given to me, in the "Overview" section we can see.
- If I will right side -> "Access the Cluster" or "Cluster UI", then I am able to see, sample code which they have given to me.
- This is something, which is UI. click and explore here & from cluster UI, itself, we can perform the all of these operation.

- So, this is my Cluster and this is my Database -> [ API Keys ] <- So, I have to send a data, I have to store a data, here inside this.
- So, How I will be able to store a data, Let's try to understand it.
- "Cluster" means set of Machines, multiple machine coming together.

API Key : eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhY2Nlc3MiOiJtIn0.GRC6VFc9Mlc9jKWSRqIixpkIvQs0m7-Idh8jTpmEG4s

In [1]:
from qdrant_client import QdrantClient

qdrant_client = QdrantClient(
    url="https://7f03220a-2f2a-4861-a0df-08ed1f679225.us-east4-0.gcp.cloud.qdrant.io:6333", 
    api_key="eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhY2Nlc3MiOiJtIn0.GRC6VFc9Mlc9jKWSRqIixpkIvQs0m7-Idh8jTpmEG4s",
)

print(qdrant_client.get_collections())

collections=[CollectionDescription(name='vijay_story')]


# Next : 

- Now, we can store the information one by one.

# Let's Install the "Qdrant Client".

- Now, next just execute the above code with my "API key", execute it.
- As we can see it outcome is [ collections=[] ] or collection is blank, which means I am able to established the communication.

In [1]:
pip install qdrant-client

  Using cached h11-0.16.0-py3-none-any.whl.metadata (8.3 kB)
  Using cached annotated_types-0.7.0-py3-none-any.whl.metadata (15 kB)
  Using cached sniffio-1.3.1-py3-none-any.whl.metadata (3.9 kB)
   ---------------------------------------- 0.0/4.6 MB ? eta -:--:--
   ---- ----------------------------------- 0.5/4.6 MB 3.4 MB/s eta 0:00:02
   --------- ------------------------------ 1.0/4.6 MB 3.6 MB/s eta 0:00:02
   --------------- ------------------------ 1.8/4.6 MB 3.2 MB/s eta 0:00:01
   ---------------------- ----------------- 2.6/4.6 MB 3.4 MB/s eta 0:00:01
   ------------------------------- -------- 3.7/4.6 MB 3.6 MB/s eta 0:00:01
   -------------------------------------- - 4.5/4.6 MB 3.7 MB/s eta 0:00:01
   ---------------------------------------- 4.6/4.6 MB 3.6 MB/s  0:00:01
Using cached h11-0.16.0-py3-none-any.whl (37 kB)
   ---------------------------------------- 0.0/2.0 MB ? eta -:--:--
   -------------------- ------------------- 1.0/2.0 MB 5.0 MB/s eta 0:00:01
   ---------

- Now, I can even print the just "Qdrant-client" variable.
- As a outcome we can see, it is giving me "Hexadecimal code". It simply means that I am able to established the communication.
- Now once we are able to established the communication.

In [2]:
qdrant_client

- Now I can store the information, that we have prepared before, into this qdrant one.
- Which is avaiable in some of the cloud platform.

# Note :

- Now, we are doing everything not in the local system, we are doing in the Global System, basically in a Cloud System.
- out data will be going to the cloud system.
- Whenever, we are going to store some sort of information, inside "Qdrant", 1st of all lets create the "Collection name".

- Variable = collection_name and I am going to give the collecation_name is "vijay_story"
- This is the colletion name which I am going to give. means a table name, So Similarily I am just trying to give a collection_name
- over here nothing much.

In [3]:
collection_name = "vijay_story"

- I am going to call my "Qdant_client", because this is my starting point, this is the point, where I will be able to communicate.
- from my local system to the "Qdant" system. 

#Note : 

- Code No.1 :
- Lets import the model 1st, as it is not able to resolved the dependencies.

- Code No.2 :
- "Qdrant_Client" I am going to call. "Qdant_Client" dot(.). I can say "recreate_collection" is a Method, which is alreay avaiable.

- Code No.3 :
- Now here, I can give the collection_name is equal to ["vijay_story"], we can try to give that is ok.

- Code No.4 :
- Now, once we are able to give a collection_name over here, we can set a vector configuration, it is nothing but the data that we are
- going to store by the way, that is something we have to give as a "Vector_configuration". Here Vector_configuation wise, lets me check
- the latest syntax, is basically ["models.VectorParameter] or Params or Paramters. 

- Code No.5 :
- Here I am just providing my vector parameters, so my vector size is basically 1536.
- After giving size basically give the ["distance = models.distance.cosine"], so whenever it is perform this operation.
- So just go-ahead with the "Cosine Similarity", This is the collection which I am going to crate.

# Note : 
- So, these syntax are already mentioned inside the documentation already so non it remebers, because there are 100 of such database,
- which is avaiable.

- So, below code this is the "collection" which I am going to create.
- This is going to craete a database kind of things, where collection name is "vijay_story" & 
- these are the parameters-> [ "vectors_config =  models.VectorParams" ] ok inside this Table or we can say collection whatever data,
- we are going to store, the size of the vector is going to be ( "1536" ) & whenever we are trying to perform, the similarity search,
- just try to use a "Cosine Similarity Search". This is what it means.
- Now, once we are able to create this collection. 
- This below code is coming from the "Qdrant" documentation itself.

In [4]:
from qdrant_client import models
qdrant_client.recreate_collection(
    collection_name = collection_name,
    vectors_config =  models.VectorParams(
        size=1536, distance = models.Distance.COSINE))

C:\Users\vijay\AppData\Local\Temp\ipykernel_28668\3436728011.py:2: DeprecationWarning: `recreate_collection` method is deprecated and will be removed in the future. Use `collection_exists` to check collection existence and `create_collection` instead.
  qdrant_client.recreate_collection(


True

# Note :

- Now once we are able to create this collection, now go back to our "Qdrant" and refresh the page and go to the "Cluster_UI &
- then clickc on -> "collection" section.
- Collections (1 / 1000)
- URL : https://7f03220a-2f2a-4861-a0df-08ed1f679225.us-east4-0.gcp.cloud.qdrant.io:6333/dashboard#/collections
- Here, we can see our collection name | Status is "Green" | Segment is 2 | Shards is 1 | Vectors-Configuration - (Name, Size, Distance)
- Name = default/ Size = 1536/ Distance = Cosine.
- Distance = Cosine - means whatever data, I am insert into this collection, it is going to follow this property. This is what it means.
- Once, I am able to create this "Collection" I can try to store our data inside this particular collection.

# Note :

- We have do the same as we did previously in "FAISS_DB". 
- So we are going to take this below data.

In [6]:
data = """ 
Making an Impact
Helping Millions of Students Succeed
Sudhanshu's commitment to affordable education wasn't just a business strategy—it was his life's mission. Over the years, iNeuron has helped over 1.5 million students from 34+ countries, providing them with the skills they need to succeed in today's competitive job market.
Many of these students, like Sudhanshu himself, came from disadvantaged backgrounds. They saw iNeuron as a lifeline—an opportunity to rise above their circumstances.
In 2022, iNeuron was acquired by PhysicsWallah in a deal worth ₹250 crore. While this acquisition was a significant milestone, Sudhanshu remained focused on his mission. Even after the acquisition, iNeuron continued to offer some of the most affordable and accessible tech courses in the world.

The Entrepreneur and Teacher: Sudhanshu's Dual Legacy
Sudhanshu's journey isn't just one of entrepreneurial success; it's also a story of dedication to teaching. Throughout his career, he has remained a passionate educator, constantly looking for ways to empower others through knowledge. Whether teaching courses in Big Data,
Data Science, or programming, Sudhanshu has always sought to make complex subjects accessible to learners at all levels.

His commitment to affordable education has earned him the respect and admiration of countless students. Many credit Sudhanshu with changing their lives, helping them secure jobs, improve their skills, and break free from the limitations of their backgrounds."""

# Next :

- Now here we are going to call a -> "data.strip()" <- This will eliminate the all those leading and trailing spaces inbetween our data. 
- Let me name it as a "Clean_data" a Variable name.
- There so many different types of clearning we can try to perform, this is just a symbolic one.
- We can perfrom a Normilization operation, we can just try to remove the pencutations, or any kind of unncessary most of the time 
- a repetitive words. So, many differen different type kind of NLP based cleaning we can try to perform on top of data, now but again that
- depends upon the kind of problem statement, which I am going to solve. So that depends more like on a business domain, or like problem
- statement that we are going to solve.
- So, as of now just "Symbolic" manner so I have called "data strip", so which going to remove leading & trailling spaces from the dataset.
- I am calling that dataset as a clean data lets suppose.
- But again, on top of this we do a lot of processing with respect to our data. 

# Next : 

- Let's suppose, I a trying to read a PDF, there will be so many different different kind of edge case, getting & I have to handle, all
- of the edge cases, then only I will be able to get the required dataset, but here dataset that we are using is not much complicated,
- There is a possibility, that I will be having some of the word or some of the line inside my dataset into a different language,
- May be Into a "Spanish", "Franch" or "Hindi" language, there will be very high chance, when I am trying to read some sort of a books,
- or maybe a PDF, or some sort of a data, there is chance that I am end of getting, mixture of a language, which we say multi lanuage.
- there is a very high possibility, I have to handle all of those situation, with respect to a dataset, then only I can try to make my  
- dataset ideal or any kind of operation, whether its a embedding or maybe whether a "RAG" or any other agentic operation. 
- alots of thing is goes around interms of building a complete data pipeline, again that is major roles and responsibility we all try to
- inside a organization.

- So, here I am able to read a data in a Symbolic manner, I am able to call a "data.Strip()". 

In [7]:
data = data.strip()

In [8]:
data

"Making an Impact\nHelping Millions of Students Succeed\nSudhanshu's commitment to affordable education wasn't just a business strategy—it was his life's mission. Over the years, iNeuron has helped over 1.5 million students from 34+ countries, providing them with the skills they need to succeed in today's competitive job market.\nMany of these students, like Sudhanshu himself, came from disadvantaged backgrounds. They saw iNeuron as a lifeline—an opportunity to rise above their circumstances.\nIn 2022, iNeuron was acquired by PhysicsWallah in a deal worth ₹250 crore. While this acquisition was a significant milestone, Sudhanshu remained focused on his mission. Even after the acquisition, iNeuron continued to offer some of the most affordable and accessible tech courses in the world.\n\nThe Entrepreneur and Teacher: Sudhanshu's Dual Legacy\nSudhanshu's journey isn't just one of entrepreneurial success; it's also a story of dedication to teaching. Throughout his career, he has remained a

# What is the total length of my string or paragraph?

- As a outcome, we can see its a -> ( 1,497 ).
- This is basically total, like a space which my dataset has occupied.

In [9]:
len(data)

1497

- Let me define some of the variables :

- Max_char or maximum character = 300, means in one single line there will be a 300 character max to max, not more then that.
- Then I can try to define "Overlapp" = 100, means lets suppose I have line No.1 & I have line No.2, 
- there is a overlapped of 100 character
- between the line No.1 and line No.2. 

- Code : [ overlap = 100 ]

- Why I am trying to keep the overlapp here ?
- Because when I am going to do a chunking operation, when I am going to breakdown this data or paragraph into a chunks.
- So, I have to maintain even a relation, means context. If there will be an overlapped, between a data, onviously it will be able to see
- the contect, it will be able to correlate.  And when I am going to do search operation, on the top of the embedding, it will be able to
- give me like, there is some match between line No.1 and there is some match between Line No.2 as well. Just to maintain that context,
- I am going to maintain the overlapping, between 2 datasets. That is the region, we have given overlapped, with hundreds of like
- a characters.

- Code : [ Chunk = []  ] <- Then try to create a chunks as a "List"
- Then I will try to keep [i] variable as a [0] zero. means start from [0] indexes. Here -> "Making an Impact" in paragraph we can see.

- Code : [ While i <len(data): ]
- while [i] is lesser (<) than, length of a (clean_data), start from [0] index go till last one, this is the region lenth of (data)
- If I have 500 word or more then that in my pagagraph it will go till last, then only it will going to break the loop,this is what I mean
- While i <len(data): <- when this statement is true, then it has to perform different different operation, 

- What kind of operation, it has to perform? 

- - it has to break down my entire dataset into a small, small pieces.

- Code : piece = data[i:i+max_char] 
- Lets suppose, here I am going to call my "data" , "data" is my variable name by the way, my whole paragraph is inside
- this variable.
- "data"[] <- inside this I am going to write the index, like start from here and then go till this particular point.
- So, I am performing slicing opeartion, I am saying start from [i] means start from [0] index, go till [i] + maximum character which is
- our 300. means [0+300]. this -> clean_data[i:i+max_char] whole will going to be one data.
- Now, lets say I will keep this data inside varaiable -> ("Piece"). This -> clean_data[i:i+max_char] <- will be having in a very 
- 1st interation, it will be having a data (i = 0th) index till 300, this is the maximum character which I have given.
- This is the line No.1 which I am going to maintain,  
- 
- Code : chunks.append(piece)
- Here, I am going to append "chunks", what ? append ("piece"). this is the list -> [ Chunks = [] ] <- which I have created.
- So, I am going to append this -> Code : piece = clean_data[i:i+max_char]  & chunks.append(piece) <- particular piece inside this
- particulat list -> [ chunks = [] ] <-

- Code : i = i+max_char - overlap 

- Now we will going to increment the value of [i], we have to shift the value of [i] now.

- Now [i] is = [i]+max_char-overlap  
- Now, I am going to reset my value of [i], where will be my [i] by the way?
- so, I has started from the [0] zero, it went till the 800, now my [i] is avaiable in the 800th location.
- But, I have to set the value of this [i] at the (700) location.why?
- So that, I will be able to maintain a overlap, of 100 inbetween. Between from the 1st sentences & between the 2nd sentences. 
- That is the region, I have written the (i+max_char) means ( 0+300 ) - 100. which means the 200, overlap.
- SO, now my [i] is here -> in 200.  [ 0,1,2,3,4,5,6.........200,201,202,203......300 ]
- so, next time, when [i] is going to start from where? this line code[i], I am talking about here -> while i<len(data):
- Next time, [i] will going to start from the 200 by the way, [200, 2001, 2002......300 ], then (200+300) = 500 technically.
- so, it will go till the (500). 
- Again, I will try to reset the value of [i] at (400). then again I will going to start from (400), then I will be able to maintain
- overlapped of 100, 100 inbetween. 
- Here, I am playing the value of [i], with those [i] those indexes, I am able to maintain the overlapped.

# Next : [ max_char = 300 ]

- Lets say, I am taking or maintain as [ max_char = 300 ].
- As a outcome, we can see over here, I am able to create 8 lines, out of my dataset & all of these lines are avaiable inside this
- particular list[].

- If we observed here, in our outcome, we can see some symbols, (\), like this we have to clean in the cleaing operation, because these
- will not going to give me value, with respect to a data. We have to perform some claning opeartion on top of this.
- There maybe some numerical value which is not of our use.
- Maybe our PDF data is avaiable in some different language, so then again we have to go with the cleaning operation. 
- There are alot of "Data preprocessing" operation, we try to perform, which is true with across all the application.
- Not just with this application, whether we are trying to build a simple "Machine Learning" model or "Deep Learning" model.
- or we are dealing with some sort of "Data analytics" everywhere, we have to go through this kind of process.

# Next, : Overlapped

- In this dataset, there are overlapped also, we we cross check here, if we go end of this data 
- -> this 1st line inside outcome, we can see -> "skills they need to succeed in toda ". ("toda") <- this only it is able to capture 
- Here, we will be able to see that 2nd paragraph of this outcome -> ["skills they need to succeed in today's competitive job market"] 
- Here, we can see full name -> Today's not "toda".
- There is overlapped, I am trying to maintain between 2 sentences, so that it will be able to maintain a contextual information.
- It will be able to understand something, like this 2 lines are connected to each others.

In [9]:
max_char = 300
overlap = 100
chunks = []
i = 0
while i <len(data):
    piece = data[i:i+max_char]
    if not piece:
        break
    chunks.append(piece)
    i = i+max_char - overlap

# Why do we need to maintain the overlapped?

- we need to maintain the overlapped, so that we will be maintain the context, a little bit of context we will be able maintain in the 
- both the sentences.
- The total is 1497, like a space which my dataset has occupied. so max to max I am able to creating 2 chunks. 
- Inside the chunks, I will be having the max to max like 2 sentences or 3 sentences.

In [10]:
chunks

["Making an Impact\nHelping Millions of Students Succeed\nSudhanshu's commitment to affordable education wasn't just a business strategy—it was his life's mission. Over the years, iNeuron has helped over 1.5 million students from 34+ countries, providing them with the skills they need to succeed in toda",
 "1.5 million students from 34+ countries, providing them with the skills they need to succeed in today's competitive job market.\nMany of these students, like Sudhanshu himself, came from disadvantaged backgrounds. They saw iNeuron as a lifeline—an opportunity to rise above their circumstances.\nIn 202",
 'backgrounds. They saw iNeuron as a lifeline—an opportunity to rise above their circumstances.\nIn 2022, iNeuron was acquired by PhysicsWallah in a deal worth ₹250 crore. While this acquisition was a significant milestone, Sudhanshu remained focused on his mission. Even after the acquisition, iNeuron ',
 "gnificant milestone, Sudhanshu remained focused on his mission. Even after the

# Check Lenth or Len

In [11]:
len(chunks)

8

# Emebedding

- I will cope and paste here exact Emebedding code previously I have used inside the "FAISS_DB".

- This is the function, -> [def generate_embeddings(text): ] <- which will going to take the data, a text and it is going 
- give the Embedding. It is going to take the data and it is going to give the Embedding.

# Next :

- Now we have the data, we have the "Chunks", from this Chunks, I have to convert this entire data into a embedding by using, this 
- "EURI API" 
- Now here this one -> [ "model": "text-embedding-3-small" ] <- we are using the OpenAI based model, again this is the multilingual model.
- that we are using, it support hundreds of languages, bydefault. When people was training this model obviously they have used the 
- multilingual dataset. 
- But, again one of the lighted & fasted embedding model and one of the low cost embedding model, we will be able to find out
- in the market. There are again hundreds of other embedding model which is avaiable. 

- For Example :
- Last class we have used from Hugging Face a "Qwen3-Embedding-8B" model. So, we can even go with that.

- So, this model -> [ "model": "text-embedding-3-small" ] will going to give us a vector size return of ( 1536 ) is the dimension of the
- one single vector, it is going to return.

In [12]:
import requests
import numpy as np

def generate_embeddings(text):
    url = "https://api.euron.one/api/v1/euri/embeddings"
    headers = {
        "Content-Type": "application/json",
        "Authorization": "Bearer euri-78c66f791ec6c9b285bc2f92d2487d5c61f0bec3c10c01a8f89fb20aba1c5ae1"
    }
    payload = {
        "input": text,
        "model": "text-embedding-3-small"
    }

    response = requests.post(url, headers=headers, json=payload)
    data = response.json()
    
    embedding = np.array(data['data'][0]['embedding'])
    
    return embedding

# Next :

- So, we have a "Chunks". we have a data, we will just run my generate embedding through this chunks, one by one & eventually,
- I will be able to get my final embedding. 

- So, here , I am going to write the code, 

- code : [ "for i in chunks": ]
- code : [ embedding = generate_embeddings(i) ] 
- code : [ print(embedding) ] or remove it, we will call it by -> [emb] <-
- one by one lets do the Chunking & call this generate embedding function, -> [ embedding = generate_embeddings(i) ] <- pass the value 
- one by one, then try to print the emdedding. If will try to keep all embedding somewhere. 
- Now lets execute below code, As a outcome we can see that, it is like creating embedding for sentences,
- for sentences No.1, Sentences No.2,Sentences No.3, Sentences No.4, Sentences No.5, Sentences No.6 etc...... 
- So, I have total 8 sentences. For all of those sentences it will able to create the embedding. So, it is able to create the numerics.

In [13]:
emb = []
for i in chunks:
    emb.append(generate_embeddings(i))

In [14]:
emb

[array([ 0.01636502, -0.01330146,  0.04360536, ..., -0.00716132,
        -0.00025957,  0.00777273], shape=(1536,)),
 array([-0.02828987, -0.00211549,  0.03156904, ..., -0.00953185,
        -0.01975844, -0.00468256], shape=(1536,)),
 array([-0.01842189, -0.03448532,  0.01168446, ..., -0.0130075 ,
         0.00758591,  0.02325387], shape=(1536,)),
 array([-0.00206819, -0.01932828,  0.04296086, ..., -0.033941  ,
         0.00859491,  0.02564766], shape=(1536,)),
 array([ 0.01036524, -0.00490744,  0.05058131, ..., -0.0313709 ,
        -0.00884518,  0.00210155], shape=(1536,)),
 array([-0.01087056, -0.02549004,  0.03719883, ..., -0.01714737,
        -0.00013056,  0.00037601], shape=(1536,)),
 array([ 0.00257804, -0.02860213,  0.04973948, ...,  0.00047777,
        -0.02138105,  0.00431918], shape=(1536,)),
 array([ 0.04722496,  0.00496422,  0.04577735, ..., -0.01690703,
        -0.07669617,  0.00453403], shape=(1536,))]

- Now above we have printed -[emb] is a list. 
- Now I have to convert this entire "List". So, that I can convert my entire data into a compatiable one.
- Code No.1 :
- So, here just variable as embedding = call numpy.array or np.array [emb] or embedding. 
- Now, yes it converted. 

In [23]:
embedding = np.array(emb)

In [24]:
embedding

array([[ 0.01636502, -0.01330146,  0.04360536, ..., -0.00716132,
        -0.00025957,  0.00777273],
       [-0.02828987, -0.00211549,  0.03156904, ..., -0.00953185,
        -0.01975844, -0.00468256],
       [-0.01842189, -0.03448532,  0.01168446, ..., -0.0130075 ,
         0.00758591,  0.02325387],
       ...,
       [-0.01087056, -0.02549004,  0.03719883, ..., -0.01714737,
        -0.00013056,  0.00037601],
       [ 0.00257804, -0.02860213,  0.04973948, ...,  0.00047777,
        -0.02138105,  0.00431918],
       [ 0.04722496,  0.00496422,  0.04577735, ..., -0.01690703,
        -0.07669617,  0.00453403]], shape=(8, 1536))

# Next : Data Dump inside a "Qdrant"

- So, our embedding is ready & our Chunk also ready. These 2 things are ready 

- Now, once my embedding is ready, Now I am good to go-ahead with my data dump inside a "Qdrant".


# Timing : 2:02:24

- Lets suppose, I am going to prepared the data.

- Code No.1 :
- Lets say Variable is "points" & List [] <- where I am going to kepp all the prepared data. 

- Code No.2 :
- Inside this particular list, I am going to keep the for idx, ( chunk,emb ) -> Enumerate ("zip,(chunks, embedding"))

- code No.3 :
- Points.append <- Means inside this list -> points [] <- try to append, what? 

- Code No.4 :
- Models.pointStruct( id=idx)
- append what? So, the models.pointstructure, try to append the (id) or id=idx. we know that for every record it is going to give me, "id"
- so, just try to generate the "id" or id=idx, then 

- code No.5 :
- then try to call emb.astype("flaot32") tolist() <- This is going to be the "Vector".

- Code No.6 :

- Payload=(text:chunk)
- Here I am trying to store even a "Meta information".
- Even last time, in "Faiss" where we have store this meta information-> ("meta_vijay.json1") & we have store this vector seperately,
- seperate file. 
- Here, I am trying to send a payroll all together. 

In [25]:
points = []

for idx, (chunk, emb) in enumerate(zip(chunks, embedding)):
    points.append(
        models.PointStruct(
            id=idx,
            vector=emb.astype("float32").tolist(),
            payload={"text": chunk}
        )
    )

# Next :

- Now, just try to check the points.
- As a outcomes, we can see this is the my data, I have end up creating, out of this -> ("PointStruct") <- 
- 1.PointStruct(id=0) <- This is the one data, & this is the vector of the data -> vector=[0.016365019604563713, -0.013301456347107887,..
- 0.007772733923047781], payload={'text': "Making an Impact\nHelping Millions of Students Succeed\nSudhanshu's commitment to affordable 

- 2.PointStruct(id=1) <- This is the one data, & this is the vector of the data -> vector=[-0.02828986570239067, -0.002115487353876233,...
- payload={'text': "1.5 million students from 34+ countries, providing them with the skills they need to succeed in today's competitive...

# Note :

- one this, if go right side to till the end of same "Vector of the data", here we can find even a "Textual information".

- Just copy below outcome and paste inside my open the "Notepad" and paste below outcome. The data is too huge.
- Just copy and paste inside my "Notepad" & just check what data, we have prepared. This is the dataset what we have prepared.
- This is the dataset, which I am going to store inside my "Vector Database" along with the textual information. Along with the 
- "Meta Information", Payload as we can see. My English Data along with this payload, along with this "Meta information".

- Above code, that we have written is nothing much we have just preparing our dataset, which is completely visiable to us.

- Now, once my dataset has been prepared. Now I am good to send my data to "Qdant" or to my Cloud Database.

In [26]:
points

[PointStruct(id=0, vector=[0.016365019604563713, -0.013301456347107887, 0.04360536113381386, 0.04943328723311424, 0.03790752589702606, -0.0390002615749836, 0.002767613623291254, 0.10979393124580383, -0.03049253113567829, 0.008149988017976284, 0.06160948425531387, -0.046961620450019836, -0.05911180377006531, -0.019291989505290985, 0.012989246286451817, -0.007050747517496347, -0.0104460334405303, -0.041732098907232285, 0.006966190412640572, 0.04503632336854935, 0.0448802188038826, 0.04329315200448036, 0.03220968693494797, -0.010055771097540855, 0.07695981860160828, -0.042876869440078735, 0.016456080600619316, 0.009249228052794933, -0.007421497255563736, -0.057862959802150726, 0.0018374872161075473, -0.01062815636396408, -0.037855491042137146, -0.009327280335128307, -0.015363344922661781, 0.010914348997175694, -0.031195005401968956, 0.0020537578966468573, -0.003444069065153599, 0.023402757942676544, 0.024820713326334953, -0.0208920668810606, 0.03655461221933365, 0.038271769881248474, -0.0

# Next :

- Here, simply I can try to call, "Qdrant_client" & I can try to call "Upsurd", inside the "Qdrant databases" that this people has 
- written, 
- Code No.2 :
- We have to mentioned the "Collection_name" which collection, we would like to "Upsurd" this data, because there is a millions of
- "Collection" that we can create for different different purposes.So, which Collection, we can "Upsurd" the data, then give me the data,
- here Points=points data that we have prepared.
- Now execute this code.

In [27]:
qdrant_client.upsert(
    collection_name=collection_name,points=points
)

UpdateResult(operation_id=0, status=<UpdateStatus.COMPLETED: 'completed'>)

# Next :

- Now it is sending my all of the data, to the "Qdrant".
- Now, If I will go back and check my Qdant Cloud, -> 
- URL : [https://7f03220a-2f2a-4861-a0df-08ed1f679225.us-east4-0.gcp.cloud.qdrant.io:6333/dashboard#/collections]
- refresh the Collection section. click on -> [ vijay_story ] enter inside my collection, we can see, we are able to find out all of my
- data points.
- All my data, it is able to store over the cloud "Qdrant".
- Once my dataset is available inside my "Qdant" , I can even do a Query.
- 1.1st is send a data or store a data.
- 2.2nd Query a data, I can try to query even here -> click on  [ vijay_story ] & click on "Points" & just type what we want to ask?
- Here it is not giving me option directly giving me an "Error". They dont have a embedding so, it will not able to generate the embedding

# Next : 

- Lets write a code here.
- lets suppose, I have a Query that, 

- query = "who is sudhanshu" <- This is the query that, I have & I have to find out the Similar kind of the record which is available
- into my "Qdrant" .
- I have to find out the similar kind of record which is available into my "Qdrant", the record which is I have already sent.
- Again same process, same procedure, we are going to follow.

In [29]:
query = "who is sudhanshu"

- 1st of all, we will try to embed this query by calling generate embedding & here I am going to pass my query.
- which is going to give me an embedding. -> (query) <- this is going to generate embedding for me.
- Lets keep inside the Variable called - ( q_resp ) or query_response
- Once my embedding being generated, once my response is avaialable.

In [30]:
q_resp = generate_embeddings(query)

In [31]:
q_resp

array([-0.02156985, -0.02583072, -0.00604044, ..., -0.00566746,
        0.01904114, -0.00011172], shape=(1536,))

- Now, what we can do is that, Numpy.array(q_resp or query_response).astype(float32).
- I will keep this inside the Variable called (q_vec) or query_vector.
- As a outcome, we can see this is the dataset which is available to me.


In [33]:
q_vec = np.array(q_resp).astype("float32")

In [34]:
q_vec

array([-0.02156985, -0.02583072, -0.00604044, ..., -0.00566746,
        0.01904115, -0.00011172], shape=(1536,), dtype=float32)

- Now I will try to call the "Qdrant_client" search <- There is a Search function, which is already given to us.
- we can do a search operation over here, to do search operation, we have to give a "Collection_name".
- Which collection we are trying to search.

- Let's suppose we are trying to search a "Sudhanshu story" or "Vijay Story", that is the "Collection_name".
- Then we are trying to send this Query vector (q_vec), which is this we are trying to send it with the search -> array([-0.02156985, -0.- 02583072, -0.00604044, ..., -0.00566746, 0.01904115, -0.00011172], shape=(1536,), dtype=float32)

- Limit is equal to (5) & Payload is equal to True <- Means whenever it is giving me response, give me a payload as well.
- Let's execute this code 

# Next :

- As a outcome, we can see that, top (5), I am looking for, ["who is Sudhanshu"] that was my query.
- I am trying to search inside my collection name is "Vijay_story", if I am looking for top (5), or top (3) data, 
- which I am sending, I am saying that ok, just give me a return with a "Payload".
- ["Payload"] means our Textual information. I can able to read as a Human Being.

- In a outcome, it is giving me the very 1st -> ["ScoredPoint(id=6)"] then ["ScoredPoint(id=4)"] then ["ScoredPoint(id=5)"] etc....
- Here, we can see the Cosine Similarity score as well. -> [ score=0.52609146, | score=0.5166283, | score=0.5161582, ] <- 
- With respect to my query -> ["Who is Sudhanshu"] <- so basically this ["ScoredPoint(id=6)"] or (id 6) is more relevant, here similary 
- score we can check it is ( 0.52609146 ) or 52 %, of the similarity score.
- & if we go right side, we can see "Text", which is readable. 

In [35]:
qdrant_client.search(collection_name="vijay_story", query_vector=q_vec, limit=5, with_payload=True)

C:\Users\vijay\AppData\Local\Temp\ipykernel_28668\3551796135.py:1: DeprecationWarning: `search` method is deprecated and will be removed in the future. Use `query_points` instead.
  qdrant_client.search(collection_name="vijay_story", query_vector=q_vec, limit=5, with_payload=True)


[ScoredPoint(id=6, version=0, score=0.52609146, payload={'text': 'accessible to learners at all levels.\n\nHis commitment to affordable education has earned him the respect and admiration of countless students. Many credit Sudhanshu with changing their lives, helping them secure jobs, improve their skills, and break free from the limitations of their backgrounds.'}, vector=None, shard_key=None, order_value=None),
 ScoredPoint(id=4, version=0, score=0.5166283, payload={'text': "reneur and Teacher: Sudhanshu's Dual Legacy\nSudhanshu's journey isn't just one of entrepreneurial success; it's also a story of dedication to teaching. Throughout his career, he has remained a passionate educator, constantly looking for ways to empower others through knowledge. Whether teaching cour"}, vector=None, shard_key=None, order_value=None),
 ScoredPoint(id=5, version=0, score=0.5161582, payload={'text': 'ate educator, constantly looking for ways to empower others through knowledge. Whether teaching cour

- So, we are able to store a data. we can even do a Top K search, on top of the data.

- This is similar to "Fiass", but here different is I am doing this is cloud, because of which I dnt have to do a setup.
- I have just taken a URL and API keys and connected and started sending a data. 
- If I have do a some sort of a query, I am able to do it as well.
- someone is managing my entire database for me.

# Public Question : we missed the Normalization ?

- Here we dont have to go with the normalization, here it is handling bydefault.
- If we go and check above code line  this one -> [ "distance = models.Distance.COSINE" ] 
- When I have created Qdrant instances, I have already written "Cosine, so it will be do a Math automatically.
- Model size is "distance" so bydefault, it will be able to do "vector(a).vector(b)/norm vector(a).norm vector(b).
- which is ideally formula of "Cosine", it will be able to do automatically.

- Incase of ["FAISS"], there is a fucntion, which is like, flat "index flat IP", it means inner product.
- The fucntion is design on that way, ok fine give me an inner product if we would like to normalized it, then normalized it.

# Public Question : Can we use something else in Distances?

- Yes many things actually, we can use in the distances, which we will discussed upcoming classes.
- Again, there are multiple way to store a data, Lets suppose my data record is in millions, so there is a concept called as ["Cluster"]
- we try to create a cluster and we try to give a centroid of the cluster and then with the help of ["Centroid"], of the cluster.
- it will try to optimize the entire, search operation. there is alot of thing we can find out interms of "Storing a data" and
- interm of performing a Search opeartion with respect to the data.